src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorSourceThumbprintAttribute.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Razor.Hosting
+{
+    /// <summary>
+    /// Specifies the hash thumprint of a source file that contributed to a compiled item.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// These attributes are added by the Razor infrastructure when generating code to assist runtime
+    /// implementations to determine the integrity of compiled items.
+    /// </para>
+    /// <para>
+    /// Runtime implementations should access the thumbprint metadata for an item using
+    /// <see cref="RazorCompiledItemExtensions.GetThumprintMetadata(RazorCompiledItem)"/>.
+    /// </para>
+    /// </remarks>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = true)]
+    public sealed class RazorSourceThumbprintAttribute : Attribute, IRazorSourceThumprintMetadata
+    {
+        /// <summary>
+        /// Creates a new <see cref="RazorSourceThumbprintAttribute"/>.
+        /// </summary>
+        /// <param name="hashAlgorithm">The hash algorithm used to create this thumbprint.</param>
+        /// <param name="thumbprint">The thumbprint as a string.</param>
+        /// <param name="kind">The kind of the source file associated with this thumbprint.</param>
+        /// <param name="identifier">The identifier associated with this thumbprint. May be <c>null</c>.</param>
+        public RazorSourceThumbprintAttribute(string hashAlgorithm, string thumbprint, string kind, string identifier)
+        {
+            if (hashAlgorithm == null)
+            {
+                throw new ArgumentNullException(nameof(hashAlgorithm));
+            }
+
+            if (thumbprint == null)
+            {
+                throw new ArgumentNullException(nameof(thumbprint));
+            }
+
+            if (kind == null)
+            {
+                throw new ArgumentNullException(nameof(kind));
+            }
+
+            HashAlgorithm = hashAlgorithm;
+            Thumbprint = thumbprint;
+            Kind = kind;
+            Identifier = identifier;
+        }
+
+        /// <summary>
+        /// Gets the hash algorithm used to create this thumbprint.
+        /// </summary>
+        public string HashAlgorithm { get; }
+        
+        /// <summary>
+        /// Gets the identifier of the source file associated with this thumbprint. May be <c>null</c>.
+        /// </summary>
+        public string Identifier { get; }
+
+        /// <summary>
+        /// Gets the kind of the source file associated with this thumbprint.
+        /// </summary>
+        public string Kind { get; }
```

Comment:
why not an enum for Kind? Same question for HashAlgorithm

Reply:
Because we don't know all of the kinds that exist. MVC defines it's own kinds.

For hashalgorithm it might be possible to use an enum, but it's more coupling. If MVC wants to use a hash algorithm that Razor doesn't know then that requires a code change in Razor. That's bad coupling.

Think of it this way, it's your data, we're just holding on to it for you 😆